### PR TITLE
Refactor component and factory interface definitions

### PR DESCRIPTION
**Pull request by tigrannajaryan**
As we are preparing for Beta release we want to cleanup publicly exported
types and interfaces and do all necessary refactoring and breaking changes
now, before the Beta release. We will have a lot less leeway for breaking
changes after the Beta release.

Component related type declarations are now all in the `component` package.
This makes it possible for the interfaces to reference each other. This
was were very restricted earlier because component interfaces were in 5
different packages and many proposals were impossible to implement because
they would result in circular dependencies between packages.

(An example upcoming new capability that is enabled by this refactoring
is for components to query the Host for other components and for factories).

List of changes in this commit:

- Move all factory interfaces and component interfaces to component package.
- Rename old factories and components interfaces to use "Old" suffix for clarity.
- Eliminate forced checks that components implement factories. This is already
  enforced by the compiler when the factory is added to the Defaults() and
  was unnecessary code.
- Eliminated some unnecessary codes (removed overall over 200 lines).
- Run `go mod tidy` on testbed.

Warning: this is a breaking change to publicly exported types and function
signatures. We announced that a breaking change is comming. Once we agree
to merge this commit we will need to announce the exact list of changes
and guide component authors to modify their components accordingly.

Future changes:
- Once all components are migrated to the new internal representation,
  delete all "Old"-suffixed definitions. This will likely be done while
  we are still in Beta phase, before the Stable release.

### CODE REVIEW

File: component/processor.go [hunk 1]
```diff
@@ -1,96 +1,123 @@
...
+
+// ProcessorFactory defines the common functions for all processor factories.
+type ProcessorFactory interface {
+	Factory
 
 	// CreateDefaultConfig creates the default configuration for the Processor.
 	// This method can be called multiple times depending on the pipeline
 	// configuration and should not cause side-effects that prevent the creation
-	// of multiple instances of the Processor.
+	// of multiple instances of the Exporter.
```
Comment (pjanotti) [inline component/processor.go]: nit: 
```suggestion
	// of multiple instances of the Processor.
```

File: component/extension.go [hunk 2]
```diff
@@ -42,3 +43,20 @@ type PipelineWatcher interface {
 	// appropriate action before that happens.
 	NotReady() error
 }
+
+// ExtensionFactory is a factory interface for extensions to the service.
+type ExtensionFactory interface {
+	Factory
+
+	// CreateDefaultConfig creates the default configuration for the Exporter.
```
Comment (pjanotti) [inline component/extension.go]:
```suggestion
	// CreateDefaultConfig creates the default configuration for the Extension.
```

File: testbed/go.sum [hunk 3]
```diff
@@ -265,8 +265,8 @@ github.com/golangci/revgrep v0.0.0-20180526074752-d9c87f5ffaf0 h1:HVfrLniijszjS1
 github.com/golangci/revgrep v0.0.0-20180526074752-d9c87f5ffaf0/go.mod h1:qOQCunEYvmd/TLamH+7LlVccLvUH5kZNhbCgTHoBbp4=
 github.com/golangci/unconvert v0.0.0-20180507085042-28b1c447d1f4 h1:zwtduBRr5SSWhqsYNgcuWO2kFlpdOZbP0+yRjmvPGys=
 github.com/golangci/unconvert v0.0.0-20180507085042-28b1c447d1f4/go.mod h1:Izgrg8RkN3rCIMLGE9CyYmU9pY2Jer6DgANEnZ/L/cQ=
-github.com/google/addlicense v0.0.0-20190510175307-22550fa7c1b0 h1:ydbHzabf84uucKri5fcfiqYxGg+rYgP/zQfLLN8lyP0=
-github.com/google/addlicense v0.0.0-20190510175307-22550fa7c1b0/go.mod h1:QtPG26W17m+OIQgE6gQ24gC1M6pUaMBAbFrTIDtwG/E=
+github.com/google/addlicense v0.0.0-20200301095109-7c013a14f2e2 h1:N/H6+jLLYNoskSbyRaph9ttbcaGEw4AWpAZw+1VqIGs=
```
Comment (pjanotti) [inline testbed/go.sum]: not sure why these changes but assuming a `go mod tidy` it is fine

**Comment by bogdandrutu** [comment]
I want to challenge this:
> Rename old factories and components interfaces to use V1 suffix for clarity.

If we break now all the users, why keeping v1 and v2, I would suggest v2 to become the version without any suffix and the old interfaces maybe name the Old or just remove them completely.

**Comment by tigrannajaryan** [comment]
> I want to challenge this:
> 
> > Rename old factories and components interfaces to use V1 suffix for clarity.
> 
> If we break now all the users, why keeping v1 and v2, I would suggest v2 to become the version without any suffix and the old interfaces maybe name the Old or just remove them completely.

Yes, that's a possibility too. Let me play with the names a bit and see what it looks like.

File: component/extension.go (hunk 2 above)
Comment (tigrannajaryan) [reply component/extension.go]: Thanks for catching.

Comment (tigrannajaryan) [reply component/extension.go]: Fixed.

File: component/processor.go (hunk 1 above)
Comment (tigrannajaryan) [reply component/processor.go]: Fixed.

File: testbed/go.sum (hunk 3 above)
Comment (tigrannajaryan) [reply testbed/go.sum]: Yes, `go mod tidy` is happy with this.

**Comment by tigrannajaryan** [comment]
@bogdandrutu I added a new commit with the approach you suggested. PTAL.

**Comment by tigrannajaryan** [comment]
What everybody thinks about the general approach of putting all factories and interfaces in one package? Do you see any downsides? Like, e.g. the potential of creating too much coupling between normally unrelated the component factories?

**Comment by tigrannajaryan** [comment]
>just remove them completely.

@bogdandrutu just to clarify, removing completely is not possible now. We have to first implemented the new interfaces in all components.

**Comment by bogdandrutu** [comment]
@tigrannajaryan the components package is very small and I would not worry that much of coupling as long as we don't put all the implementations there :)

Few questions/suggestions:
1. I am curious why consumer is outside, maybe that should be moved as well.
2. `data` is too generic for the new internal data package we should probably fix that before making it public.

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan the components package is very small and I would not worry that much of coupling as long as we don't put all the implementations there :)
> 
> Few questions/suggestions:
> 
>     1. I am curious why consumer is outside, maybe that should be moved as well.
> 

I'd keep this as a separate package. We will move new internal data representation implementation here once it is ready, so it will become a large package itself. I don't want to overload `component` package too much.

These are data models, they don't need any access to component factories / interfaces and I think it is better to keep them in a separate package. It is a typical approach to keep data model declarations as a separate package so that other packages can freely import them in any order without worrying about circular dependencies.

>     2. `data` is too generic for the new internal data package we should probably fix that before making it public.

Yes, these will need to be moved/renamed. But I want to wait until you finish your work and we have the implementation.

**Comment by tigrannajaryan** [comment]
@bogdandrutu any more thoughts on this?

**Comment by bogdandrutu** [comment]
@tigrannajaryan for the moment I am done, will look again once it is merge (in some examples where we use this and see how it is).

File: component/exporter.go [hunk 4]
```diff
@@ -1,32 +1,60 @@
-// Copyright 2019, OpenTelemetry Authors
+// Copyright  OpenTelemetry Authors
```
Comment (jrcamp) [inline component/exporter.go]: nit: Is this supposed to have the year? Happened in a couple places.
